src/main/java/com/getindata/connectors/http/sink/httpclient/JavaNetSinkHttpClient.java

```diff
@@ -0,0 +1,111 @@
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of {@link SinkHttpClient} that uses Java 11's {@link HttpClient}.
+ */
+public class JavaNetSinkHttpClient implements SinkHttpClient {
+  private final HttpClient httpClient;
+
+  public JavaNetSinkHttpClient() {
+    this.httpClient = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build();
+  }
+
+  /**
+   * A wrapper structure around an arbitrary element, keeping a reference to a particular
+   * {@link HttpSinkRequestEntry}. Used internally by the {@code HttpSinkWriter} to pass
+   * {@code HttpSinkRequestEntry} along some other element that is logically connected with it
+   * (e.g., full HTTP request built from the {@code HttpSinkRequestEntry}).
+   *
+   * @param <T>
+   */
+  @RequiredArgsConstructor
+  @EqualsAndHashCode
+  private static class HttpSinkRequestEntryWrapper<T> {
+    /**
+     * An element logically connected with the {@link HttpSinkRequestEntry}.
+     */
+    public final T element;
+
+    /**
+     * A representation of a single {@link HttpSink} request.
+     */
+    public final HttpSinkRequestEntry sinkRequestEntry;
+  }
+
+  @Override
+  public CompletableFuture<SinkHttpClientResponse> putRequests(
+      List<HttpSinkRequestEntry> requestEntries, String endpointUrl
+  ) {
+    var endpointUri = URI.create(endpointUrl);
+
+    List<HttpSinkRequestEntryWrapper<HttpRequest>> requests = requestEntries
+        .stream()
+        .map(requestEntry -> new HttpSinkRequestEntryWrapper<>(
+            buildHttpRequest(requestEntry, endpointUri),
+            requestEntry
+        ))
+        .collect(Collectors.toList());
+
+    return getCompletedFutures(requests).thenApply(this::prepareHttpClientResponse);
+  }
+
+  private HttpRequest buildHttpRequest(HttpSinkRequestEntry requestEntry, URI endpointUri) {
+    return HttpRequest
+        .newBuilder()
+        .uri(endpointUri)
+        .version(java.net.http.HttpClient.Version.HTTP_1_1)
+        .header("Content-Type", requestEntry.contentType)
+        .method(requestEntry.method, HttpRequest.BodyPublishers.ofByteArray(requestEntry.element))
+        .build();
+  }
+
+  private CompletableFuture<List<HttpSinkRequestEntryWrapper<Optional<HttpResponse<String>>>>> getCompletedFutures(
+      List<HttpSinkRequestEntryWrapper<HttpRequest>> requests
+  ) {
+    var futures = requests
+        .stream()
+        .map(req -> httpClient
+            .sendAsync(req.element, HttpResponse.BodyHandlers.ofString())
+            .exceptionally(ex -> null)
+            .thenApply(res -> new HttpSinkRequestEntryWrapper<>(Optional.ofNullable(res), req.sinkRequestEntry))
```

Comment:
Using Optionals as arguments is considered as a bad practice or at least debatable.
Optional should be used as return types, not arguments.

I would say that `HttpSinkRequestEntryWrapper` could return Optional<T> in getElement instead accepting it as an argument.